### PR TITLE
[Diffusion][MiniMax-H3] Add SM90 Sage compute for SubBlock sparse attention

### PR DESCRIPTION
**Pull request by IPostYellow**
## Motivation
Inspired by the combination of block-sparse attention and Sage FP8 compute in [FlashInfer's SM100 backend](https://github.com/flashinfer-ai/flashinfer/pull/4612), this PR adds optional SageAttention2 compute to MiniMax-H3's SubBlock sparse attention backend on SM90.

The new mode is opt-in. BF16 remains the default, and Sage compute currently targets SM90.

## Modifications

- Add `compute_mode="sage_fp8"` with online INT8-QK / FP8-PV computation and BF16 inputs and outputs. Model weights remain unchanged.
- Adapt SubBlock routing to the native SM90 kernel's Q64 × K128 block layout.
- Preserve the existing BF16 Q64 × K64 defaults on SM90, SM100, and SM120.
- Check GPU capability and the optional SpargeAttention dependency at server startup.
- Add regression tests and documentation.

Enable Sage compute with:

```bash
--attention-backend subblock_sparse_attn \
--attention-backend-config '{"compute_mode":"sage_fp8"}'
```

## Accuracy Tests

- Native SM90 kernel tests: **2 passed**, covering full-budget attention with a ragged tail and sparse attention with variable block counts.
- MiniMax-H3 end-to-end tests: **81/81 completed** — 3 methods × 3 cases (T2VA, FL2VA, REF2VA-V) × 3 durations (5/10/15 seconds) × 3 seeds (0/1/2).
- Video SSIM is measured against the matching dense FA output. All frames are included, with YUV420 weighting `(4 × Y + U + V) / 6`; results are averaged across three seeds.
- `ruff check`, `ruff format --check`, and `git diff --check` passed.

### Video SSIM

Dense FA is the reference (SSIM = 1). SB denotes SubBlock BF16.

| Case | Duration | SB | SB+Sage |
|---|---:|---:|---:|
| T2VA | 5s | 0.821579 | 0.818747 |
| T2VA | 10s | 0.870746 | 0.865564 |
| T2VA | 15s | 0.891421 | 0.884516 |
| FL2VA | 5s | 0.851103 | 0.841919 |
| FL2VA | 10s | 0.778914 | 0.777729 |
| FL2VA | 15s | 0.815655 | 0.814308 |
| REF2VA-V | 5s | 0.943797 | 0.939980 |
| REF2VA-V | 10s | 0.980822 | 0.979227 |
| REF2VA-V | 15s | 0.963598 | 0.959587 |

### Audio similarity

Decoded audio is compared against Dense FA for the same case, duration, and seed. Values are means across three seeds, in dB. Higher waveform SNR and lower log-mel MAE indicate closer signals.

The original 32 kHz stereo audio is retained. Log-mel MAE uses 80 mel bands, a 2048-sample FFT, a 512-sample hop, and a shared floor 80 dB below the reference mel peak.

| Case | Duration | SB SNR ↑ | SB+Sage SNR ↑ | SB log-mel MAE ↓ | SB+Sage log-mel MAE ↓ |
|---|---:|---:|---:|---:|---:|
| T2VA | 5s | 12.450 | 11.507 | 2.006 | 1.989 |
| T2VA | 10s | 13.332 | 12.356 | 1.693 | 1.761 |
| T2VA | 15s | 13.146 | 11.232 | 1.733 | 1.810 |
| FL2VA | 5s | 11.090 | 11.617 | 2.353 | 2.281 |
| FL2VA | 10s | 10.627 | 11.074 | 2.358 | 2.113 |
| FL2VA | 15s | 6.215 | 5.885 | 2.468 | 2.612 |
| REF2VA-V | 5s | 15.819 | 13.038 | 0.552 | 0.916 |
| REF2VA-V | 10s | 22.119 | 19.429 | 0.435 | 0.586 |
| REF2VA-V | 15s | 14.882 | 11.927 | 1.245 | 1.528 |

Comparing SB+Sage directly against SB, group-mean SNR ranges from 15.08 to 20.78 dB and log-mel MAE from 0.484 to 1.386 dB.

## Speed Tests and Profiling

Benchmarked MiniMax-H3 on 8 H200 GPUs with identical inputs and generation settings across methods. Sparse settings: `sparsity=0.75`, `n_q=n_k=4`, `skip_first_steps=10`. The benchmark explicitly sets `n_k=4`; Sage's default is `n_k=8`.

### Generation time

Time is the mean server-side generation time across three seeds, excluding warmup. The ↓ columns report time reduction relative to Dense FA: `(dense baseline time - method time) / dense baseline time`.

| Case | Duration | Dense FA (s) | SB (s) | SB ↓ | SB+Sage (s) | SB+Sage ↓ |
|---|---:|---:|---:|---:|---:|---:|
| T2VA | 5s | 39.21 | 30.38 | 22.52% | 28.46 | 27.42% |
| T2VA | 10s | 111.84 | 78.16 | 30.11% | 67.90 | 39.29% |
| T2VA | 15s | 218.72 | 144.44 | 33.96% | 121.35 | 44.52% |
| FL2VA | 5s | 42.51 | 33.55 | 21.09% | 30.92 | 27.26% |
| FL2VA | 10s | 116.30 | 81.78 | 29.68% | 70.92 | 39.02% |
| FL2VA | 15s | 224.70 | 149.94 | 33.27% | 126.07 | 43.89% |
| REF2VA-V | 5s | 134.83 | 94.13 | 30.19% | 80.98 | 39.94% |
| REF2VA-V | 10s | 421.82 | 271.58 | 35.62% | 219.48 | 47.97% |
| REF2VA-V | 15s | 842.06 | 520.66 | 38.17% | 411.83 | 51.09% |

### Peak reserved memory

Measured by the PyTorch allocator on the output rank. Values are the maximum per-request peak across three seeds, in GiB.

| Case | Duration | Dense FA | SB | SB+Sage |
|---|---:|---:|---:|---:|
| T2VA | 5s | 83.51 | 83.51 | 83.51 |
| T2VA | 10s | 88.30 | 88.30 | 88.30 |
| T2VA | 15s | 93.10 | 93.10 | 93.28 |
| FL2VA | 5s | 84.42 | 84.42 | 84.42 |
| FL2VA | 10s | 88.64 | 88.64 | 88.77 |
| FL2VA | 15s | 93.54 | 93.54 | 93.54 |
| REF2VA-V | 5s | 84.90 | 84.90 | 84.90 |
| REF2VA-V | 10s | 92.71 | 92.71 | 92.71 |
| REF2VA-V | 15s | 99.23 | 99.23 | 99.23 |

Sage was faster than BF16 SubBlock in all 27 matched-seed comparisons. Group-mean generation time decreases by **6.32%–20.90%**, with an SSIM decrease of **0.00118–0.00918** relative to BF16 SubBlock. Peak reserved memory remains essentially unchanged.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #34293968939](https://github.com/sgl-project/sglang/actions/runs/34293968939)
Latest PR Test (Extra): :x: [Run #34293968672](https://github.com/sgl-project/sglang/actions/runs/34293968672)
Latest PR Test (AMD ROCm 7.2): :x: [Run #34293968907](https://github.com/sgl-project/sglang/actions/runs/34293968907)

### CODE REVIEW

**Review by BBuf — APPROVED** [review]

File: test/registered/kernel/attention/test_subblock_sage_fp8_sm90.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+import math
+import unittest
+
+import torch
+
+from sglang.test.ci.ci_register import register_cuda_ci
+from sglang.test.test_utils import CustomTestCase
+
+register_cuda_ci(est_time=45, stage="base-b-kernel-unit", runner_config="1-gpu-large")
```
Comment (mickqian) [inline test/registered/kernel/attention/test_subblock_sage_fp8_sm90.py:12]: [P2] Move the new test into the supported registered-test taxonomy. The [failed lint job](https://github.com/sgl-project/sglang/actions/runs/34242724603/job/102116742998) rejects this newly added `test/registered/kernels/...` path: `scripts/lint/check_registered_tests.py` requires `test/registered/<kind>/<subsystem>/`, and the CUDA kernel kind is singular `kernel`, not `kernels`. Move this file to e.g. `test/registered/kernel/attention/test_subblock_sage_fp8_sm90.py` and update its path reference in the SubBlock README. The existing `stage="base-b-kernel-unit"` registration is compatible with the `kernel` category and should be preserved. Rerunning the unchanged head will hit the same lint failure and block the JIT jobs again.

**Review by mickqian — COMMENTED** [review]
CI follow-up for 634692134872: the lint failure is deterministic; both failed JIT unit-test jobs stopped at `Fast-fail: lint check failed`, before running tests. Please fix the registered-test path below, then run `pre-commit run check-registered-tests --files test/registered/kernel/attention/test_subblock_sage_fp8_sm90.py` and rerun CI. No kernel regression is established by these failed jobs.

File: test/registered/kernel/attention/test_subblock_sage_fp8_sm90.py (hunk 1 above)
Comment (IPostYellow) [reply test/registered/kernel/attention/test_subblock_sage_fp8_sm90.py:12]: Thanks for catching this. Fixed in d0e3c4c369.

Moved the test to `test/registered/kernel/attention/test_subblock_sage_fp8_sm90.py` and updated the README reference.

**Comment by BBuf** [comment]
https://github.com/sgl-project/sglang/actions/runs/34293968939/job/102286669006?pr=37982
